crates/transaction-generator-lib/src/workflow_delegator.rs

```diff
@@ -0,0 +1,257 @@
+/// We do this until pool_0 is empty.
+///
+/// We proceed, until in the last stage - stage n - calls gen_n, which consumes accounts from pool_n-1.
+///
+/// There are two modes on when to move to the next stage:
+/// - WhenDone means as soon as pool_i is empty, we move to stage i+1
+/// - ExternallySet means we wait for external signal to move to next stage, and we stop creating transactions
+///   until we receive it (or will move early if pool hasn't been consumed yet)
+///
+/// Use WorkflowTxnGeneratorCreator::create_workload to create this generator.
+struct WorkflowTxnGenerator {
+    stage: StageTracking,
+    generators: Vec<Box<dyn TransactionGenerator>>,
+    pool_per_stage: Vec<Arc<ObjectPool<LocalAccount>>>,
+    num_for_first_stage: usize,
+    // Internal counter, so multiple workers (WorkflowTxnGenerator) can coordinate how many times to execute the first stage
+    completed_for_first_stage: Arc<AtomicUsize>,
+}
+
+impl WorkflowTxnGenerator {
+    fn new(
+        stage: StageTracking,
+        generators: Vec<Box<dyn TransactionGenerator>>,
+        pool_per_stage: Vec<Arc<ObjectPool<LocalAccount>>>,
+        num_for_first_stage: usize,
+        completed_for_first_stage: Arc<AtomicUsize>,
+    ) -> Self {
+        Self {
+            stage,
+            generators,
+            pool_per_stage,
+            num_for_first_stage,
+            completed_for_first_stage,
+        }
+    }
+}
+
+impl TransactionGenerator for WorkflowTxnGenerator {
+    fn generate_transactions(
+        &mut self,
+        account: &LocalAccount,
+        mut num_to_create: usize,
+    ) -> Vec<SignedTransaction> {
+        assert_ne!(num_to_create, 0);
+        let mut stage = self.stage.load_current_stage();
+
+        match &self.stage {
+            StageTracking::WhenDone(stage_counter) => {
+                if stage == 0 {
+                    let prev = self
+                        .completed_for_first_stage
+                        .fetch_add(num_to_create, Ordering::Relaxed);
+                    num_to_create =
+                        cmp::min(num_to_create, self.num_for_first_stage.saturating_sub(prev));
+
+                    if num_to_create == 0 {
+                        info!("TransactionGenerator Workflow: Stage 0 is full with {} accounts, moving to stage 1", self.pool_per_stage.get(0).unwrap().len());
+                        let _ = stage_counter.compare_exchange(
+                            0,
+                            1,
+                            Ordering::Relaxed,
+                            Ordering::Relaxed,
+                        );
+                        stage = 1;
+                    }
+                } else if stage < self.pool_per_stage.len()
+                    && self.pool_per_stage.get(stage - 1).unwrap().len() == 0
+                {
+                    info!("TransactionGenerator Workflow: Stage {} has consumed all accounts, moving to stage {}", stage, stage + 1);
```

Comment:
Won't multiple threads enter this part ? In that case won't we have to adjust num_to_create accordingly ? Or rather I miss the consumption from the pool part

Reply:
yes, multiple will enter. 

but since generators themselves depend on the pool, and grab elements from the pool, when pool is empty, even if we pass num_to_create, they will only create as many as they could grab from the pool. so the pool itself is the coordinator

added a comment above